### PR TITLE
Gradient descent function

### PR DESCRIPTION
**Pull request by conor012**
Uses gradient descent to find global minimum for functions y = x^2 and y = x^2 + x^4 + 15 in as many dimensions as you want.

The learning rate, maximum number of iterations, initial point and tolerances must be set.

The tolerances come in the form of the the L2 norm between in the gradient \nabla F(x) at step x_i and x_{i-1} and the relative difference in the objective value between step x_i and x_{i-1}.

The algorithm will continue to move to the closer to the global minimum as long as the maximum number of iterations has not been reached an neither tolerance has been reached.

I hope that I've done this pull request correctly. I will add some notes for this algorithm tomorrow.

### CODE REVIEW

File: Gradient Descent/gradient.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+const double epsilon = pow(10,-6); // Fix a small epsilon for finite difference scheme to calculate grad
+
+// This is the function to be minimised. At the moment it is f(x) = x^2
+double objfun(vector<double>& x){               // Takes a vector x...
+    double output_val = 0.0;
+    for(int i=0 ; i < x.size(); i++){
+        output_val += pow(x[i],2);}             //... and computes x_i^2 for each coordinate...
+    return output_val;                      //... and returns vector of calculation.
+}
```
Comment (Tom271) [inline Gradient Descent/gradient.cpp:17]: True in 1D, actually calculating x^T x (inner product) in higher dimensions (\sum_{i=1}^N x_i^2 0 for x \in R^N

File: Gradient Descent/gradient.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+// This another function to be minimised. Uncomment out if
+// you want to use it. At the moment it is f(x) = x^2 + x^4 + 15
+//double objfun(vector<double>& x){               // Takes a vector x...
+//    double output_val = 0.0;
+//    for(int i=0 ; i < x.size(); i++){
+//        output_val += pow(x[i],2) + pow(x[i],4);}             //... and computes x_i^2 +x_i^4 + 15 for each coordinate...
+//    return output_val + 15;                      //... and returns vector of calculation.
```
Comment (Tom271) [inline Gradient Descent/gradient.cpp:25]: Again this returns a double, so is the sum of the function evaluated at each point. 

This isn't wrong, just need to be careful in our definitions of the functions as they map from R^N to R

File: Gradient Descent/gradient.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+// Function to multiply every element in a vector x by a scalar k
+vector<double> multiplication_vectorscal(vector<double> &x, double k){
+    for (int i = 0; i < x.size() ; i++)
+    {
+        x[i] *= k;
+    }
+    return x;
+}
```
Comment (Tom271) [inline Gradient Descent/gradient.cpp:81]: All these helper functions can be put in a separate header file -- I think most will be handled by Eigen

File: Gradient Descent/gradient.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+int main(){
+    double gam = 0.0001; // Set the learning rate
+    int n_iter = 100000; // Set the maximum number of iterations
+    double grad_change_tol = pow(10,-10); // Set the tolerance for change of gradient
+    double rel_sol_change_tol = pow(10,-10); // Set the tolernace for change of solution
+    
+    // Initialise vectors to hold the minimum value to be found and the intial values (which need to be inputed)
+    vector<double> min;
```
Comment (Tom271) [inline Gradient Descent/gradient.cpp:138]: min is a std function, try to avoid using it as a variable name

File: Gradient Descent/gradient.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    vector<double> int_vals;
+
+    // Set intial values. This could be anything. Have chosen x_i = 2^i for i=1,2,3,4.
+    for (int i = 0; i < 4; i++)
```
Comment (Tom271) [inline Gradient Descent/gradient.cpp:142]: Is this the only place the dimension is set?

**Review by Tom271 — APPROVED** [review]
Good work! Thanks for all the comments, the code makes sense. I'm working on translating this into the structure I've come up with, I'll have that sorted by this evening

File: Gradient Descent/gradient.cpp (hunk 1 above)
Comment (conor012) [reply Gradient Descent/gradient.cpp:17]: Of course! That's quite silly of me

File: Gradient Descent/gradient.cpp (hunk 5 above)
Comment (conor012) [reply Gradient Descent/gradient.cpp:142]: Yes this is the only place where the dimension is set at the moment
